nixos/tests/matrix-appservice-irc.nix

```diff
@@ -0,0 +1,155 @@
+  in
+  {
+
+    name = "matrix-synapse";
+    meta = {
+      maintainers = pkgs.matrix-appservice-irc.meta.maintainers;
+    };
+
+    nodes = {
+      homeserver = { pkgs, ... }: {
+        # We'll switch to this once the config is copied into place
+        specialisation.running.configuration = {
+          services.matrix-synapse = {
+            enable = true;
+            database_type = "sqlite3";
+            app_service_config_files = [ "/registration.yml" ];
+
+            enable_registration = true;
+
+            listeners = [
+              # The default but tls=false
+              { "bind_address" = ""; "port" = 8448; "resources" = [{ "compress" = true; "names" = [ "client" "webclient" ]; } { "compress" = false; "names" = [ "federation" ]; }]; "tls" = false; "type" = "http"; "x_forwarded" = false; }
+            ];
+          };
+
+          networking.firewall.allowedTCPPorts = [ 8448 ];
+        };
+      };
+
+      ircd = { pkgs, ... }: {
+        services.ngircd = {
+          enable = true;
+          config = ''
+            [Global]
+              Name = ircd.ircd
+              Info = Server Info Text
+              AdminInfo1 = _
+
+            [Channel]
+              Name = #test
+              Topic = a cool place
+
+            [Options]
+              PAM = no
+          '';
+        };
+        networking.firewall.allowedTCPPorts = [ 6667 ];
+      };
+
+      appservice = { pkgs, ... }: {
+        services.matrix-appservice-irc = {
+          enable = true;
+          registrationUrl = "http://appservice:8009";
+
+          settings = {
+            homeserver.url = homeserverUrl;
+            homeserver.domain = "homeserver";
+
+            ircService.servers."ircd" = {
+              name = "IRCd";
+              port = 6667;
+              dynamicChannels = {
+                enabled = true;
+                aliasTemplate = "#irc_$CHANNEL";
+              };
+            };
+          };
+        };
+
+        networking.firewall.allowedTCPPorts = [ 8009 ];
+      };
+
+      client = { pkgs, ... }: {
+        environment.systemPackages = [
+          (pkgs.writers.writePython3Bin "do_test"
+            { libraries = [ pkgs.python3Packages.irc pkgs.python3Packages.matrix-client ]; } ''
+            import socket
+            from matrix_client.client import MatrixClient
+            from time import sleep
+
+            m_client = MatrixClient("${homeserverUrl}")
+            m_client.register_with_password(username="alice", password="foobar")
+
+            i_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            i_sock.connect(("ircd", 6667))
+            i_sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+            i_sock.send(b"USER bob bob bob :bob\n")
+            i_sock.send(b"NICK bob\n")
+
+            m_room = m_client.join_room("#irc_#test:homeserver")
+            i_sock.send(b"JOIN #test\n")
+
+            # plenty of time for the joins to happen
+            sleep(10)
```

Comment:
This could be improved by parsing the proper events. For IRC you could wait for the topic.

Reply:
I don't think it's that simple. In both directions, not only does my client have to join, but the bridged user *also* has to join, to be able to see the coming message.

If we had one user wait for the other's bridge-user, it'd be a complex matrix of racey possiblities.

Reply:
A maybe-nicer alternative though: spawn two threads each sending a message once per second to the respective server. Even if the first few are missed, it doesn't matter, eventually the right one will come through.

Reply:
This is fine for now if it's not racy.